### PR TITLE
[Compression V2] Movement pruning

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###
Implement movement pruning in this paper.
https://arxiv.org/abs/2005.07683

### Checklist ###
  - [x] test case
  - [x] doc

### CODE REVIEW

File: docs/en_US/Compression/v2_pruning_algo.rst [hunk 1]
```diff
@@ -273,6 +274,48 @@ User configuration for ADMM Pruner
 
 .. autoclass:: nni.algorithms.compression.v2.pytorch.pruning.ADMMPruner
 
+Movement Pruner
+---------------
+
+Movement pruner is an implementation of movement pruning.
+This is a pruning by step algorithm, the masks may change during each step.
```
Comment (QuanluZhang) [inline docs/en_US/Compression/v2_pruning_algo.rst:281]: what is "step algorithm"?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    StraightMetricsCalculator,
+    NormalSparsityAllocator
+)
+
+_logger = logging.getLogger(__name__)
+
+
+class PrunerScoredModuleWrapper(Module):
+    """
+    Wrap an module to enable data parallel, forward method customization and buffer registeration.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:29]: an -> a

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        torch.nn.init.constant_(self.weight_score, val=0.0)
+
+        # register buffer for mask
+        self.register_buffer("weight_mask", torch.ones(self.module.weight.shape))
+        if hasattr(self.module, 'bias') and self.module.bias is not None:
+            self.register_buffer("bias_mask", torch.ones(self.module.bias.shape))
+            self.bias = Parameter(torch.empty(self.module.bias.size()))
+            self.bias.data = self.module.bias.data
+        else:
+            self.register_buffer("bias_mask", None)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:65]: if there is no bias, do we need to create bias_mask?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    def _reset_optimizer(self):
+        """
+        Weed out the weight_score from the parameters passed to optimizer, guaranteed to load the optimizer state dict.
+        """
+        if self._origin_optimizer_cls is not None:
+            optimizer_grouped_parameters = [{
+                "params": [p for n, p in self.compressor.bound_model.named_parameters() if "weight_score" not in n and p.requires_grad]
+            }]
+            if self._origin_optimizer_cls.__name__ == 'SGD':
+                self.optimizer = self._origin_optimizer_cls(optimizer_grouped_parameters, lr=0.001)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:125]: why hardcode 0.001 here?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        The total `optimizer.step()` number before start pruning for warm up.
+        Make sure `warm_up_step` is smaller than `cool_down_beginning_step`.
+    cool_down_beginning_step: int
+        The number of steps at which sparsity stops growing, note that the sparsity stop growing doesn't mean masks not changed.
+        The sparsity after each `optimizer.step()` is:
+        total_sparsity * (1 - (1 - (current_step - warm_up_step) / (cool_down_beginning_step - warm_up_step)) ** 3).
+    """
+    def __init__(self, model: Module, config_list: List[Dict], trainer: Callable[[Module, Optimizer, Callable], None],
+                 optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor], training_epochs: int, warm_up_step: int,
+                 cool_down_beginning_step: int):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:192]: is it better to provide default value for some parameters?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        assert self.warm_up_step < self.cool_down_beginning_step, '`warm_up_step` should smaller than `cool_down_beginning_step`'
+        super().__init__(model, config_list)
+
+    def _validate_config_before_canonical(self, model: Module, config_list: List[Dict]):
+        schema_list = [deepcopy(NORMAL_SCHEMA), deepcopy(EXCLUDE_SCHEMA), deepcopy(INTERNAL_SCHEMA)]
+        schema = CompressorSchema(schema_list, model, _logger)
+        schema.validate(config_list)
+
+    def cubic_schedule(self, current_step: int):
+        if self.warm_up_step < current_step <= self.cool_down_beginning_step:
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:208]: python supports this style?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            self.sparsity_allocator = NormalSparsityAllocator(self, continuous_mask=False)
+
+        # use Adam to update the weight_score
+        params = [{"params": [p for n, p in self.bound_model.named_parameters() if "weight_score" in n and p.requires_grad]}]
+        optimizer = Adam(params, 1e-2)
+        self.step_counter = 0
+
+        # update the masks after each optimzier step
+        def _optimizer_patch():
+            optimizer.step()
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:228]: no need to call optimizer.step()?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 8]
```diff
@@ -133,7 +133,8 @@ def patched_criterion(output, target):
         super().__init__(compressor)
         self.trainer = trainer
         self.training_epochs = training_epochs
-        self._origin_optimizer = optimizer
+        self._origin_optimizer_cls = optimizer.__class__ if optimizer is not None else None
+        self._origin_optimizer_state_dict = optimizer.state_dict() if optimizer is not None else None
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:137]: what cases is this update not able to handle?

Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:137]: some fancy usage of optimizer?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 9]
```diff
@@ -173,6 +164,18 @@ def reset(self):
         self._hook_buffer = {}
         self._add_all_hook()
 
+    def _reset_optimizer(self):
+        self.compressor._unwrap_model()
+        if self._origin_optimizer_cls is not None:
+            if self._origin_optimizer_cls.__name__ == 'SGD':
+                self.optimizer = self._origin_optimizer_cls(self.compressor.bound_model.parameters(), lr=0.001)
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:171]: lr is hardcoded 0.001?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py [hunk 10]
```diff
@@ -339,6 +342,8 @@ def __init__(self, pruner: Compressor, dim: Optional[Union[int, List[int]]] = No
             Example:
 
             The metric size is (12,), and block_sparse_size=[64], then the mask will expand to (768,) at first before expand with `dim`.
+        continuous_mask
+            Inherit the mask already in the wrapper if set True.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:346]: what is this for?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py [hunk 11]
```diff
@@ -9,7 +9,15 @@
 from .base import MetricsCalculator
 
 __all__ = ['NormMetricsCalculator', 'MultiDataNormMetricsCalculator', 'DistMetricsCalculator',
-           'APoZRankMetricsCalculator', 'MeanRankMetricsCalculator']
+           'APoZRankMetricsCalculator', 'MeanRankMetricsCalculator', 'StraightMetricsCalculator']
+
+
+class StraightMetricsCalculator(MetricsCalculator):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py:15]: better to add docstring for it

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 10 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:346]: whether continuous depends on specific sparsity allocator, is it proper to put this argument in base sparsity allocator?

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 1 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst:281]: I want to say `pruning by step` algorithm... means this pruner will generate and apply masks during each optimizer step.

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:125]: Because lr is a required parameter in `SGD`, and `0.001` is a fake number, it will be reload by `self.optimizer.load_state_dict(self._origin_optimizer_state_dict)`

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 3 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:65]: In fact, we don't need it, but we want to be compatible with the old version.

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 5 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:192]: we can't provide default values because we don't know how many steps are in an epoch.
or we can provide, `warm_up_step=None`, `cool_down_beginning_step=None`, meaning it won't warm up and cool down?

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 6 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:208]: yes

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 7 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py:228]: it is needed, this optimizer only handles and updates the `weight_score` in wrapper.
`_optimizer_patch` will be called after each `patched_optimzier.step()`

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 8 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:137]: just don't want to reference the parameters in model...

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 9 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:171]: same as previous

File: nni/algorithms/compression/v2/pytorch/pruning/tools/base.py (hunk 10 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/base.py:346]: In fact, each sparsity allocator can have `continuous_mask`, so I put this in base,

File: nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py (hunk 11 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/metrics_calculator.py:15]: add it

File: nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py (hunk 2 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py]: fix it

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 1 above)
Comment (zheng-ningxin) [reply docs/en_US/Compression/v2_pruning_algo.rst:281]: I think every pruner is pruning step by step? What's the concrete meaning of `step` here?

Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst:281]: oh yes, it is easy to be misunderstood here, this means after each `optimizer.step()`, the model will be applied a new mask. I will update the docstring later.

File: docs/en_US/Compression/v2_pruning_algo.rst [hunk 12]
```diff
@@ -273,6 +274,65 @@ User configuration for ADMM Pruner
...
+.. code-block:: python
+
+   from nni.algorithms.compression.v2.pytorch.pruning import MovementPruner
+   config_list = [{'op_types': ['Linear'], 'op_partial_names': ['bert.encoder'], 'sparsity': 0.9}]
+   pruner = MovementPruner(model, config_list, p_trainer, optimizer, criterion, 10, 3000, 27000)
+   masked_model, masks = pruner.compress()
+
+   # ignore the parameters with `weight_score` in name if you want to finetune with masks
+   optimizer_grouped_parameters = [{
+        "params": [p for n, p in model.named_parameters() if "weight_score" not in n and p.requires_grad]
```
Comment (zheng-ningxin) [inline docs/en_US/Compression/v2_pruning_algo.rst:306]: What is `weight_score` for? Can we handle this automatically so that user don't have to modify the optimizer manually? Besides, whether `weight_score`  limits our appliable scenario to a specific implement version/repo of transformer?

Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst:306]: image

`weight_score` is register in wrapper as a parameter, it is the sum of `- weight * weight_grad`.
It's OK that user directly use `optimizer = Adam(model.named_parameters(), lr=2e-5)`, just some computing resources were wasted. But it's a good idea that we handle this automatically, I will try this.

`weight_score` will not limit our appliable scenario, all module that has `weight` can use this pruner.

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 1 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst]: change to
`This is a "fine-pruning" algorithm, which means the masks may change during each fine-tuning step.`

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 12 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst]: > Can we handle this automatically so that user don't have to modify the optimizer manually?

fix it

**Review by QuanluZhang — APPROVED** [review]

**Review by zheng-ningxin — APPROVED** [review]
